engine/function/src/templates/Numeric.ftl

```diff
@@ -2314,21 +2553,32 @@ public class Numeric {
             while (vi.hasNext()) {
                 final ${pt.primitive} c = vi.${pt.iteratorNext}();
                 final ${pt2.primitive} w = wi.${pt2.iteratorNext}();
+
+               <#if pt.valueType.isFloat >
                 if (isNaN(c)) {
                     return Double.NaN;
                 }
+               </#if>
+
+               <#if pt2.valueType.isFloat >
                 if (isNaN(w)) {
                     return Double.NaN;
                 }
+               </#if>
 
                 if (!isNull(c) && !isNull(w)) {
-                    vsum += c * w;
+                   <#if pt.valueType.isFloat >
+                    vsum += (double) c * w;
+                   <#else>
+                    vsum += c * (double) w;
+                   </#if>
                 }
             }
         }
 
         return vsum;
```

Comment:
Do we need some infinity handling here?

Reply:
Better short-circuiting was added. Infinity handling should be fine.

Reply:
Unit test added.